packages/langium/src/references/references.ts

```diff
@@ -71,15 +77,239 @@ export class DefaultReferences implements References {
+        const refs: ReferenceDescription[] = [];
+        if (includeDeclaration) {
+            const ref = this.getReferenceToSelf(target);
+            if (ref) {
+                refs.push(ref);
+            }
+        }
+        const localReferences: Reference[] = [];
+        streamAst(rootNode).forEach(node => {
+            streamReferences(node).forEach(refInfo => {
+                if (refInfo.reference.ref === target) {
+                    localReferences.push(refInfo.reference);
+                }
+            });
+        });
+        localReferences.forEach(ref => {
+            refs.push({
+                sourceUri: getDocument(ref.$refNode.element).uri,
+                sourcePath: this.nodeLocator.getAstNodePath(ref.$refNode.element),
+                targetUri: getDocument(target).uri,
+                targetPath: this.nodeLocator.getAstNodePath(target),
+                segment: toDocumentSegment(ref.$refNode),
+                local: equalURI(getDocument(ref.$refNode.element).uri, getDocument(target).uri)
+            });
+        });
+        return stream(refs);
+    }
+
+    findAllGlobalReferences(target: AstNode, includeDeclaration: boolean): Stream<ReferenceDescription> {
+        const refs: ReferenceDescription[] = [];
+        if (includeDeclaration) {
+            const ref = this.getReferenceToSelf(target);
+            if (ref) {
+                refs.push(ref);
+            }
+        }
+        refs.push(...this.index.findAllReferences(target, this.nodeLocator.getAstNodePath(target)));
+        return stream(refs);
+    }
+
+    findLocalReferencesToTypeAttribute(target: TypeAttribute, rootNode: AstNode): Stream<ReferenceDescription> {
+        const refs: ReferenceDescription[] = [];
+        const interfaceNode = getContainerOfType(target, isInterface);
+        if (interfaceNode) {
+            const interfaces = collectChildrenTypes(interfaceNode, this, this.documents, this.nodeLocator);
+            const targetRules: Array<ParserRule | Action> = [];
+            interfaces.forEach(interf => {
+                const rules = this.findLocalRulesWithReturnType(interf, rootNode);
+                targetRules.push(...rules);
+            });
+            if (equalURI(getDocument(target).uri, getDocument(rootNode).uri)) {
+                const ref = this.getReferenceToSelf(target);
+                if (ref) {
+                    refs.push(ref);
+                }
+            }
+            targetRules.forEach(rule => {
+                const assignment = isParserRule(rule) ?
+                    extractAssignments(rule.definition).find(a => a.feature === target.name)
+                    : getContainerOfType(rule, isGroup)!.elements.find(el => isAssignment(el) && el.feature === target.name);
+                if (assignment?.$cstNode) {
+                    const leaf = findLeafNodeAtOffset(assignment.$cstNode!, assignment.$cstNode!.offset);
```

Comment:
Suggestion: Using the `$cstNode` of the whole assignment to get the offset of the assigned property isn't recommended here. Instead, you should find the correct node by calling `findNodeForFeature(assignment.$cstNode, 'feature')`.
```suggestion
                    const leaf = findNodeForFeature(assignment.$cstNode, 'feature');
```